ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,435 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            return STATUS_INSUFFICIENT_RESOURCES;
+        }
+
+        /* Fill it by munged prefix and our previously created device string */
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &MungedPrefix);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &DeviceString);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        DPRINT("Munged string is '%wZ'\n", &MungedStringReal);
+    }
+
+    DPRINT("GUID string is '%wZ'\n", &GuidStringReal);
+    DPRINT("Reference string is '%wZ'\n", &ReferenceStringReal);
+
+    /* Store received parts if the parameters are not null */
+    if (PrefixString != NULL)
+    {
+        if (RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE))
+        {
+            PrefixString->Buffer = SymbolicLinkName->Buffer;
+            PrefixString->Length = KernelModePrefix.Length;
+            PrefixString->MaximumLength = PrefixString->Length;
+        }
+        else if (RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
+        {
+            PrefixString->Buffer = SymbolicLinkName->Buffer;
+            PrefixString->Length = UserModePrefix.Length;
+            PrefixString->MaximumLength = PrefixString->Length;
+        }
+
+        DPRINT("Prefix string is '%wZ'\n", PrefixString);
+    }
+
+    if (MungedString != NULL)
+        *MungedString = MungedStringReal;
     else
-        InstanceKeyRealP = &InstanceKeyReal;
+        RtlFreeUnicodeString(&MungedStringReal);
 
-    *GuidKeyRealP = NULL;
-    *DeviceKeyRealP = NULL;
-    *InstanceKeyRealP = NULL;
+    if (GuidString != NULL)
+        *GuidString = GuidStringReal;
 
-    RtlInitUnicodeString(&BaseKeyU, BaseKeyString);
+    if (ReferenceString != NULL)
+        *ReferenceString = ReferenceStringReal;
 
-    /* Open the DeviceClasses key */
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &BaseKeyU,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               NULL,
-                               NULL);
-    Status = ZwOpenKey(&ClassesKey,
-                       DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                       &ObjectAttributes);
+    if (ReferenceStringPresent != NULL)
+        *ReferenceStringPresent = ReferenceStringReal.Length > sizeof(WCHAR);
+
+    if (InterfaceClassGuid != NULL)
+    {
+        /* Convert GUID string into a GUID and store it also */
+        Status = RtlGUIDFromString(&GuidStringReal, InterfaceClassGuid);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlGUIDFromString() failed with status 0x%08lx\n", Status);
+            RtlFreeUnicodeString(&MungedStringReal);
+            RtlInitEmptyUnicodeString(MungedString, NULL, 0);
+            return Status;
+        }
+    }
+
+    /* We're done */
+    return STATUS_SUCCESS;
```

Comment:
You might as well, explicitly set `Status` to `STATUS_SUCCESS` just before taking the `if (InterfaceClassGuid != NULL)` case (or, leave it as it is, since in principle, when reaching that `if`-block, its value should be success), and then here just do a `return Status;` and removing the one that is above (line 222), as it's going to fall back there and do the same action.

Reply:
Improved, thanks. :slightly_smiling_face: